### DISCUSSION TITLE
Uhrzeit stimmt nicht immer.

### DISCUSSION THREAD

**Discussion by fraeggle** [beabel/radiowecker#49] [Fragen & Antworten / Q&A] [answered]
Ich werde wahnsinnig.
Ich habe die Online Version 3.0.2 genommen. Ich habe das Repository von hier genommen und mit Arduino compiliert und geflashed.
Jedesmal das gleiche Ergebnis. Mal stimmt die Uhrzeit, in der nächsten Minute läuft sie tatsächlich 20 oder mehr Sekunden länger, dann kann es sein Sie stimmt wieder. Auch egal, ob ich die Fritzbox als Zeitserver nehme oder de.pool.ntp.org. Hab ich jedoch eine Verbindung über die Webseite sieht es aus als würde die Zeitaktualisierung richtig funktionieren. Beende ich die WEB-Session habe ich fast sofort wieder das oben genannte Problem. Selbst wenn das Webradio an ist passt das mit der Uhrzeit nicht. Nur wenn ich über den Browser mit dem Radiowecker verbunden bin geht die Zeit richtig.
Überseh ich was???
![Radio_Version](https://github.com/beabel/radiowecker/assets/40692505/7cd58878-115d-4184-9205-e9ddc30b8a08)
Wenn ich so die Verbindung stehen lassen geht es:
![Web_Verbindung](https://github.com/beabel/radiowecker/assets/40692505/825abfc4-5029-4e7c-a9c3-a49147685282)

**Comment by fraeggle** [comment]
Wenn ich den 
  //timed event updatetime display every minute
  if ((millis() - tick) > 60000)  auf 30000 setze, 
passiert der "Fehler" nicht. Dann stimmt die Uhrzeit immer.

**Comment by 2010-Measure** [comment]
Hallo,
Das ist alles so richtig. 
Die Zeitaktualisierung wird über millis vorgenommen und dies ist nicht synchron mit der RTC. 
Abhilfe kann man nur schaffen, wenn man dies über die Uhrzeit steuert. 
Ich mache die Zeitaktualisierung alle 15s, dann legt der Fehler immer zwischen 0 und 15s, genauer brauche ich es nicht.
Besser ist es, den Sekundenoffset zu berücksichtigen, wie es im Beitrag **_**Versatz der Uhrzeit**_** beschrieben wurde.

**Comment by ChrisH172** [comment]
Hallo,

die Zeile if ((millis() - tick) > 60000) ist schon eine gute Lösung, allerdings ist die nachfolgende Zeile mit der Berechnung von tick an dieser Stelle falsch, da die Zeitstruktur ti nicht aktuell ist. Dadurch entstehen diese sehr ungenauen Sprünge in der Minute. Wenn man die tick-Berechnung in die Funktion updateTime() nach getLocalTime(&ti) verlegt, läuft die Uhr einwandfrei. Auch muß tick in setup() nicht mehr  explizit initialisiert werden, da dies showClock() indirekt über updateTime() übernimmt. Im Anhang habe ich mal die beiden geänderten Files auf Basis der Version 3.0.5 abgelegt.

Die Veränderung der Abfrage von 60000 auf 30000 oder 15000 (s. vorherige Kommentare) bringt zwar eine Verbesserung, aber halbiert bzw. viertelt die eingestellte Einschlafzeit.
[tft_display.txt](https://github.com/user-attachments/files/17422555/tft_display.txt)
[radiowecker.txt](https://github.com/user-attachments/files/17422556/radiowecker.txt)

**Comment by beabel** [comment, marked as answer]
Das Problem sollte ab Version 3.0.6 gelöst sein.

Thx an @ChrisH172

**Reply by fraeggle** [reply]
Hast du da seine Änderung reingemacht?

**Reply by fraeggle** [reply]
Auf jeden Fall tuts mit einen 2 Dateien... Super Arbeit von euch 2...........  :-)

**Reply by beabel** [reply]
Ich habe folgendes geändert:
[Changes](https://github.com/beabel/radiowecker/commit/23280b2a7b0a9239ad3b5e2084cfe9c120cb6ea5)

**Comment by fraeggle** [comment]
Definitiv erledigt. Suuuuper Arbeit von euch....
